cli/tools/bundle/mod.rs

```diff
@@ -0,0 +1,879 @@
+
+  {
+    tokio::spawn(async move {
+      loop {
+        tokio::select! {
+            res = esbuild.wait_for_exit() => {
+                eprintln!("esbuild exited: {:?}", res);
+                break;
+            }
+        }
+      }
+
+      Ok::<(), AnyError>(())
+    });
+  }
+
+  bundle_flags.external.push("*.node".into());
+
+  let mut builder = EsbuildFlagsBuilder::default();
+  builder
+    .bundle(bundle_flags.bundle)
+    .minify(bundle_flags.minify)
+    .splitting(bundle_flags.code_splitting)
+    .external(bundle_flags.external.clone())
+    .format(match bundle_flags.format {
+      BundleFormat::Esm => esbuild_rs::Format::Esm,
+      BundleFormat::Cjs => esbuild_rs::Format::Cjs,
+      BundleFormat::Iife => esbuild_rs::Format::Iife,
+    })
+    .packages(match bundle_flags.packages {
+      PackageHandling::External => esbuild_rs::PackagesHandling::External,
+      PackageHandling::Bundle => esbuild_rs::PackagesHandling::Bundle,
+    });
+  if let Some(outdir) = bundle_flags.output_dir.clone() {
+    builder.outdir(outdir);
+  } else {
+    builder.outfile(
+      bundle_flags
+        .output_path
+        .clone()
+        .unwrap_or_else(|| "./dist/bundled.js".to_string()),
+    );
+  }
+  let flags = builder.build().unwrap();
+
+  let entries = roots.into_iter().map(|e| ("".into(), e.into())).collect();
+
+  let response = client
+    .send_build_request(protocol::BuildRequest {
+      entries,
+      key: 0,
+      flags: flags.to_flags(),
+      write: true,
+      stdin_contents: None.into(),
+      stdin_resolve_dir: None.into(),
+      abs_working_dir: init_cwd.to_string_lossy().to_string(),
+      context: false,
+      mangle_cache: None.into(),
+      node_paths: vec![],
+      plugins: Some(vec![protocol::BuildPlugin {
+        name: "deno".into(),
+        on_start: false,
+        on_end: false,
+        on_resolve: (vec![protocol::OnResolveSetupOptions {
+          id: 0,
+          filter: ".*".into(),
+          namespace: "".into(),
+        }]),
+        on_load: vec![protocol::OnLoadSetupOptions {
+          id: 0,
+          filter: ".*".into(),
+          namespace: "".into(),
+        }],
+      }]),
+    })
+    .await
+    .unwrap();
+
+  if response.errors.len() > 0 {
+    for error in &response.errors {
+      eprintln!(
+        "{}: {}",
+        deno_terminal::colors::red("bundler error"),
+        format_message(error)
+      );
+    }
+  }
+
+  if response.warnings.len() > 0 {
+    for warning in &response.warnings {
+      eprintln!(
+        "{}: {}",
+        deno_terminal::colors::yellow("bundler warning"),
+        format_message(warning)
+      );
+    }
+  }
+
+  if let Some(stdout) = response.write_to_stdout {
+    println!("{}", String::from_utf8_lossy(&stdout));
+  } else if response.errors.len() == 0 {
+    if bundle_flags.output_dir.is_none()
+      && std::env::var("NO_DENO_BUNDLE_HACK").is_err()
+    {
+      // TODO(nathanwhit): MASSIVE HACK
```

Comment:
Maybe move it to a helper function and add some comments on the purpose of this hack?

Are there tests that cover it?

Reply:
Extracted it out and added a test/comments